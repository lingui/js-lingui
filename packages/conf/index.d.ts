import type { GeneratorOptions } from "@babel/core"

export declare type CatalogFormat =
  | "lingui"
  | "minimal"
  | "po"
  | "csv"
  | "po-gettext"
export type CatalogFormatOptions = {
  origins?: boolean
  lineNumbers?: boolean
}
export declare type OrderBy = "messageId" | "origin"
declare type CatalogConfig = {
  name?: string
  path: string
  include: string[]
  exclude?: string[]
}

export type LocaleObject = {
  [locale: string]: string[] | string
}

export type DefaultLocaleObject = {
  default: string
}

export declare type FallbackLocales = LocaleObject | DefaultLocaleObject

declare type CatalogService = {
  name: string
  apiKey: string
}

declare type ExtractorType = {
  match(filename: string): boolean
  extract(filename: string, targetDir: string, options?: any): void
}

export declare type LinguiConfig = {
  catalogs: CatalogConfig[]
  compileNamespace: "es" | "cjs" | "ts" | string
  extractBabelOptions: Record<string, unknown>
  compilerBabelOptions: GeneratorOptions
  fallbackLocales: FallbackLocales
  format: CatalogFormat
  extractors?: ExtractorType[]
  prevFormat: CatalogFormat
  formatOptions: CatalogFormatOptions
  localeDir: string
  locales: string[]
  catalogsMergePath: string
  orderBy: OrderBy
  pseudoLocale: string
  rootDir: string
  runtimeConfigModule:
    | [source: string, identifier?: string]
    | {
        i18n?: [source: string, identifier?: string]
        Trans?: [source: string, identifier?: string]
      }
  sourceLocale: string
  service: CatalogService
}
export declare const defaultConfig: LinguiConfig
export declare function getConfig({
  cwd,
  configPath,
  skipValidation,
}?: {
  cwd?: string
  configPath?: string
  skipValidation?: boolean
}): LinguiConfig

export declare function makeConfig(
  userConfig: Partial<LinguiConfig>,
  opts?: {
    skipValidation?: boolean
  }
): LinguiConfig

export declare const configValidation: {
  exampleConfig: {
    extractBabelOptions: {
      extends: string
      rootMode: string
      plugins: string[]
      presets: string[]
    }
    compilerBabelOptions: GeneratorOptions
    catalogs: CatalogConfig[]
    compileNamespace: "es" | "ts" | "cjs" | string
    fallbackLocales: FallbackLocales
    format: CatalogFormat
    formatOptions: CatalogFormatOptions
    locales: string[]
    orderBy: OrderBy
    pseudoLocale: string
    rootDir: string
    runtimeConfigModule: LinguiConfig["runtimeConfigModule"]
    sourceLocale: string
    service: CatalogService
  }
  deprecatedConfig: {
    fallbackLocale: (
      config: LinguiConfig & DeprecatedFallbackLanguage
    ) => string
    localeDir: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathDirs: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathIgnorePatterns: (
      config: LinguiConfig & DeprecatedLocaleDir
    ) => string
  }
  comment: string
}
export declare function replaceRootDir(
  config: LinguiConfig,
  rootDir: string
): LinguiConfig
/**
 * Replace fallbackLocale with fallbackLocales
 *
 * Released in lingui-conf 0.9
 * Remove anytime after 4.x
 */
declare type DeprecatedFallbackLanguage = {
  fallbackLocale: string | null
}

export declare function fallbackLanguageMigration(
  config: LinguiConfig & DeprecatedFallbackLanguage
): LinguiConfig
/**
 * Replace localeDir, srcPathDirs and srcPathIgnorePatterns with catalogs
 *
 * Released in @lingui/conf 3.0
 * Remove anytime after 4.x
 */
declare type DeprecatedLocaleDir = {
  localeDir: string
  srcPathDirs: string[]
  srcPathIgnorePatterns: string[]
}
export declare function catalogMigration(
  config: LinguiConfig & DeprecatedLocaleDir
): LinguiConfig
export {}
//# sourceMappingURL=index.d.ts.map
