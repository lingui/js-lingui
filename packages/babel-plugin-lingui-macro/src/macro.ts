import { createMacro, MacroParams } from "babel-plugin-macros"

import type { VisitNodeObject } from "@babel/traverse"
import { Program } from "@babel/types"

import linguiPlugin from "./index"
import { JsMacroName, JsxMacroName } from "./constants"

function macro({ state, babel, config }: MacroParams) {
  if (!state.get("linguiProcessed")) {
    state.opts = config
    const plugin = linguiPlugin(babel)

    const { enter, exit } = plugin.visitor.Program as VisitNodeObject<
      any,
      Program
    >

    enter(state.file.path, state)
    state.file.path.traverse(plugin.visitor, state)
    exit(state.file.path, state)

    state.set("linguiProcessed", true)
  }

  return { keepImports: true }
}

;[
  JsMacroName.defineMessage,
  JsMacroName.msg,
  JsMacroName.t,
  JsMacroName.useLingui,
  JsMacroName.plural,
  JsMacroName.select,
  JsMacroName.selectOrdinal,

  JsxMacroName.Trans,
  JsxMacroName.Plural,
  JsxMacroName.Select,
  JsxMacroName.SelectOrdinal,
].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throw new Error(
        `The macro you imported from "@lingui/core/macro" or "@lingui/react/macro" is being executed outside the context of compilation with babel-plugin-macros. ` +
          `This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. ` +
          `Please see the documentation for how to configure babel-plugin-macros properly: ` +
          "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md"
      )
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
}) as { isBabelMacro: true }
