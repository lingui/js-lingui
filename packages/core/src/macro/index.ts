import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJs } from "./macroJs"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export type LinguiMacroOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

export const jsMacroTags = new Set([
  "defineMessage",
  "arg",
  "t",
  "plural",
  "select",
  "selectOrdinal",
])

let config: LinguiConfigNormalized

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { i18nImportModule, i18nImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsNodes = new Set<NodePath>()
  let needsI18nImport = false

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsMacroTags.has(tagName)) {
      nodes.forEach((node) => {
        jsNodes.add(node.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsNodesArray = Array.from(jsNodes)

  jsNodesArray.filter(isRootPath(jsNodesArray)).forEach((path) => {
    const macro = new MacroJs(babel, { i18nImportName, stripNonEssentialProps })
    if (macro.replacePath(path)) needsI18nImport = true
  })

  if (needsI18nImport) {
    addImport(
      babel.types,
      state.file.path.node,
      i18nImportModule,
      i18nImportName
    )
  }
}

;[...jsMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/core/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
