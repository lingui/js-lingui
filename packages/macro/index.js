module.exports = require("@lingui/babel-plugin-lingui-macro/macro")
