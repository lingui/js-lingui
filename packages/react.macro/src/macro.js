import { createMacro } from "babel-plugin-macros"
import { implementationPlugin as TransformReactPlugin } from "@lingui/babel-plugin-transform-react"

const importsToCarryOver = ["DateFormat", "NumberFormat"]

function macro({ references, state, babel }) {
  const { types: t } = babel
  const plugin = new TransformReactPlugin(babel)

  const toKeepImports = ["Trans"]

  // Trick the plugin into thinking we've processed an import
  plugin.setImportDeclarations(
    Object.keys(references).reduce((obj, key) => {
      if (key !== "default" || key !== "Trans") obj[key] = key
      return obj
    }, {})
  )

  for (let [tagName, tags] of Object.entries(references)) {
    if (importsToCarryOver.includes(tagName)) toKeepImports.push(tagName)

    tags.forEach(openingTag => {
      if (!t.isJSXOpeningElement(openingTag.container)) return // Exclude closing elements

      const node = openingTag.context.parentPath.container

      plugin.JSXElement({ node }, state.file)
    })
  }

  const linguiReactImport = state.file.path.node.body.find(
    importNode =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === "@lingui/react"
  )

  // Handle adding the import or altering the existing import
  if (linguiReactImport) {
    toKeepImports.forEach(name => {
      if (
        linguiReactImport.specifiers.findIndex(
          specifier => specifier.imported && specifier.imported.name === name
        ) === -1
      ) {
        linguiReactImport.specifiers.push(
          t.importSpecifier(t.identifier(name), t.identifier(name))
        )
      }
    })
  } else {
    state.file.path.node.body.unshift(
      t.importDeclaration(
        toKeepImports.map(name =>
          t.importSpecifier(t.identifier(name), t.identifier(name))
        ),
        t.stringLiteral("@lingui/react")
      )
    )
  }
}

export default createMacro(macro)
// For intellisense as these will be ignored by babel-macro
// Might want to copy PropTypes here?
export const Trans = () => {}
export const Plural = () => {}
export const Select = () => {}
export const SelectOrdinal = () => {}
export const DateFormat = () => {}
export const NumberFormat = () => {}
