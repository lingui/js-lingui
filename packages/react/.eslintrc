{
  "extends": ["../../.eslintrc", "plugin:react/recommended", "plugin:react-hooks/recommended"],
  "settings": {
    "react": {
      "version": "detect" // "detect" automatically picks the version of React that is installed
    }
  },
  "rules": {
    "react/prop-types": "off",
    "react/display-name": "off"
  }
}
