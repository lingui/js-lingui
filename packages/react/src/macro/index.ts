import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJSX } from "./macroJsx"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  LinguiMacroOpts,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export const jsxMacroTags = new Set([
  "Trans",
  "Plural",
  "Select",
  "SelectOrdinal",
])

let config: LinguiConfigNormalized

// todo, make memoization on the conf library level
function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { TransImportModule, TransImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsxNodes = new Set<NodePath>()

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsxMacroTags.has(tagName)) {
      // babel-plugin-macros return JSXIdentifier nodes.
      // Which is for every JSX element would be presented twice (opening / close)
      // Here we're taking JSXElement and dedupe it.
      nodes.forEach((node) => {
        // identifier.openingElement.jsxElement
        jsxNodes.add(node.parentPath.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsxNodesArray = Array.from(jsxNodes)

  jsxNodesArray.filter(isRootPath(jsxNodesArray)).forEach((path) => {
    const macro = new MacroJSX(babel, { stripNonEssentialProps })
    macro.replacePath(path)
  })

  if (jsxNodes.size) {
    addImport(
      babel.types,
      state.file.path.node,
      TransImportModule,
      TransImportName
    )
  }
}

;[...jsxMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/react/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
