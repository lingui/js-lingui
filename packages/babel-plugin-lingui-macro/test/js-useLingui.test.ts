import { makeConfig } from "@lingui/conf"
import { macroTester } from "./macroTester"

describe.skip("", () => {})

macroTester({
  cases: [
    {
      name: "tagged template literal style",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}
    `,
    },
    {
      name: "support renamed destructuring",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t: _ } = useLingui();
  const a = _\`Text\`;
}
    `,
    },
    {
      name: "should process macro with  matching name in correct scopes",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;

  {
    // here is child scope with own "t" binding, shouldn't be processed
    const t = () => {};
    t\`Text\`;
  }
  {
    // here is child scope which should be processed, since 't' relates to outer scope
    t\`Text\`;
  }
}
    `,
    },
    {
      name: "inserted statement should not clash with existing variables",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const _t = "i'm here";
  const { t: _ } = useLingui();
  const a = _\`Text\`;
}
    `,
    },
    {
      name: "support nested macro",
      code: `
import { useLingui } from '@lingui/react/macro';
import { plural } from '@lingui/core/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text \${plural(users.length, {
          offset: 1,
          0: "No books",
          1: "1 book",
          other: "# books"
        })}\`;
}

    `,
    },
    {
      name: "support message descriptor",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t({ message: "Hello", context: "my custom" });
}
    `,
    },
    {
      name: "support passing t variable as dependency",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = useMemo(() => t\`Text\`, [t]);
}
    `,
    },
    {
      name: "transform to standard useLingui statement",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { i18n, t } = useLingui();

  console.log(i18n);
  const a = t\`Text\`;
}
    `,
    },
    {
      name: "work with existing useLingui statement",
      code: `
import { useLingui as useLinguiMacro } from '@lingui/react/macro';
import { useLingui } from '@lingui/react';

function MyComponent() {
  const { _ } = useLingui();

  console.log(_);
  const { t } = useLinguiMacro();
  const a = t\`Text\`;
}
    `,
    },

    {
      name: "work with renamed existing useLingui statement",
      code: `
import { useLingui as useLinguiRenamed  } from '@lingui/react';
import { useLingui as useLinguiMacro } from '@lingui/react/macro';

function MyComponent() {
  const { _ } = useLinguiRenamed();

  console.log(_);
  const { t } = useLinguiMacro();
  const a = t\`Text\`;
}
    `,
    },
    {
      name: "should not break on function currying",
      code: `
  import { useLingui } from '@lingui/core/macro';

  const result = curryingFoo()()
  console.log('curryingFoo', result)
    `,
    },
    {
      name: "work with multiple react components",
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}

function MyComponent2() {
  const { t } = useLingui();
  const b = t\`Text\`;
}`,
    },
    {
      name: "support configuring runtime module import using LinguiConfig.runtimeConfigModule",
      macroOpts: {
        linguiConfig: makeConfig(
          {
            runtimeConfigModule: {
              useLingui: ["@my/lingui-react", "myUselingui"],
            },
          },
          { skipValidation: true }
        ),
      },
      code: `
import { useLingui } from '@lingui/react/macro';

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}
`,
    },
  ],
})
