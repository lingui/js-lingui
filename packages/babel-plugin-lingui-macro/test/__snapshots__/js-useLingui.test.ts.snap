// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inserted statement should not clash with existing variables 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const _t = "i'm here";
  const { t: _ } = useLingui();
  const a = _\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const _t = "i'm here";
  const { _: _t2 } = _useLingui();
  const a = _t2(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`should not break on function currying 1`] = `
import { useLingui } from "@lingui/core/macro";
const result = curryingFoo()();
console.log("curryingFoo", result);

↓ ↓ ↓ ↓ ↓ ↓

const result = curryingFoo()();
console.log("curryingFoo", result);

`;

exports[`should process macro with  matching name in correct scopes 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;

  {
    // here is child scope with own "t" binding, shouldn't be processed
    const t = () => {};
    t\`Text\`;
  }
  {
    // here is child scope which should be processed, since 't' relates to outer scope
    t\`Text\`;
  }
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
  {
    // here is child scope with own "t" binding, shouldn't be processed
    const t = () => {};
    t\`Text\`;
  }
  {
    // here is child scope which should be processed, since 't' relates to outer scope
    _t(
      /*i18n*/
      {
        id: "xeiujy",
        message: "Text",
      }
    );
  }
}

`;

exports[`support configuring runtime module import using LinguiConfig.runtimeConfigModule 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { myUselingui as _useLingui } from "@my/lingui-react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`support message descriptor 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = t({ message: "Hello", context: "my custom" });
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "BYqAaU",
      message: "Hello",
      context: "my custom",
    }
  );
}

`;

exports[`support nested macro 1`] = `
import { useLingui } from "@lingui/react/macro";
import { plural } from "@lingui/core/macro";

function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text \${plural(users.length, {
    offset: 1,
    0: "No books",
    1: "1 book",
    other: "# books",
  })}\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "hJRCh6",
      message:
        "Text {0, plural, offset:1 =0 {No books} =1 {1 book} other {# books}}",
      values: {
        0: users.length,
      },
    }
  );
}

`;

exports[`support passing t variable as dependency 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = useMemo(() => t\`Text\`, [t]);
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = useMemo(
    () =>
      _t(
        /*i18n*/
        {
          id: "xeiujy",
          message: "Text",
        }
      ),
    [_t]
  );
}

`;

exports[`support renamed destructuring 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t: _ } = useLingui();
  const a = _\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`tagged template literal style 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`transform to standard useLingui statement 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { i18n, t } = useLingui();

  console.log(i18n);
  const a = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { i18n, _: _t } = _useLingui();
  console.log(i18n);
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`work with existing useLingui statement 1`] = `
import { useLingui as useLinguiMacro } from "@lingui/react/macro";
import { useLingui } from "@lingui/react";

function MyComponent() {
  const { _ } = useLingui();

  console.log(_);
  const { t } = useLinguiMacro();
  const a = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
import { useLingui } from "@lingui/react";
function MyComponent() {
  const { _ } = useLingui();
  console.log(_);
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`work with multiple react components 1`] = `
import { useLingui } from "@lingui/react/macro";
function MyComponent() {
  const { t } = useLingui();
  const a = t\`Text\`;
}

function MyComponent2() {
  const { t } = useLingui();
  const b = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}
function MyComponent2() {
  const { _: _t2 } = _useLingui();
  const b = _t2(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;

exports[`work with renamed existing useLingui statement 1`] = `
import { useLingui as useLinguiRenamed } from "@lingui/react";
import { useLingui as useLinguiMacro } from "@lingui/react/macro";

function MyComponent() {
  const { _ } = useLinguiRenamed();

  console.log(_);
  const { t } = useLinguiMacro();
  const a = t\`Text\`;
}

↓ ↓ ↓ ↓ ↓ ↓

import { useLingui as useLinguiRenamed } from "@lingui/react";
import { useLingui as _useLingui } from "@lingui/react";
function MyComponent() {
  const { _ } = useLinguiRenamed();
  console.log(_);
  const { _: _t } = _useLingui();
  const a = _t(
    /*i18n*/
    {
      id: "xeiujy",
      message: "Text",
    }
  );
}

`;
